src/core/lib/event_engine/windows/windows_listener.cc

```diff
@@ -0,0 +1,356 @@
+
+WindowsEventEngineListener::~WindowsEventEngineListener() {
+  GRPC_EVENT_ENGINE_TRACE(
+      "%s",
+      absl::StrFormat("WindowsEventEngineListener::%p shutting down", this)
+          .c_str());
+  // Shut down each port listener before destroying this EventEngine::Listener
+  for (auto& port_listener : port_listeners_) {
+    port_listener.reset();
+  }
+  on_shutdown_(absl::OkStatus());
+}
+
+absl::StatusOr<int> WindowsEventEngineListener::Bind(
+    const EventEngine::ResolvedAddress& addr) {
+  if (started_.load()) {
+    return absl::FailedPreconditionError(
+        absl::StrFormat("WindowsEventEngineListener::%p is already started, "
+                        "ports can no longer be bound",
+                        this));
+  }
+  int out_port = ResolvedAddressGetPort(addr);
+  EventEngine::ResolvedAddress out_addr(addr);
+  EventEngine::ResolvedAddress tmp_addr;
+  // Check if this is a  wildcard port, and if so, try to keep the port the same
+  // as some previously created listener.
+  if (out_port == 0) {
+    grpc_core::MutexLock lock(&socket_listeners_mu_);
+    for (const auto& port_listener : port_listeners_) {
+      tmp_addr = port_listener->listener_sockname();
+      out_port = ResolvedAddressGetPort(tmp_addr);
+      if (out_port > 0) {
+        ResolvedAddressSetPort(out_addr, out_port);
+        break;
+      }
+    }
+  }
+  if (ResolvedAddressToV4Mapped(out_addr, &tmp_addr)) {
+    out_addr = tmp_addr;
+  }
+  // Treat :: or 0.0.0.0 as a family-agnostic wildcard.
+  if (ResolvedAddressIsWildcard(out_addr)) {
+    out_addr = ResolvedAddressMakeWild6(out_port);
+  }
+  // open the socket
+  SOCKET sock = WSASocket(AF_INET6, SOCK_STREAM, IPPROTO_TCP, nullptr, 0,
+                          IOCP::GetDefaultSocketFlags());
+  if (sock == INVALID_SOCKET) {
+    auto error = GRPC_WSA_ERROR(WSAGetLastError(), "WSASocket");
+    return GRPC_ERROR_CREATE_REFERENCING("Failed to add port to server", &error,
+                                         1);
+  }
+  auto port_listener = AddSinglePortSocketListener(sock, out_addr);
+  GRPC_RETURN_IF_ERROR(port_listener.status());
+  return (*port_listener)->port();
+}
+
+absl::Status WindowsEventEngineListener::Start() {
+  GPR_ASSERT(!started_.exchange(true));
+  grpc_core::MutexLock lock(&socket_listeners_mu_);
+  for (auto& port_listener : port_listeners_) {
+    GRPC_RETURN_IF_ERROR(port_listener->Start());
+  }
+  return absl::OkStatus();
+}
+
+absl::StatusOr<WindowsEventEngineListener::SinglePortSocketListener*>
+WindowsEventEngineListener::AddSinglePortSocketListener(
+    SOCKET sock, EventEngine::ResolvedAddress addr) {
+  auto single_port_listener =
+      SinglePortSocketListener::Create(this, sock, addr);
+  GRPC_RETURN_IF_ERROR(single_port_listener.status());
+  auto* single_port_listener_ptr = single_port_listener->get();
+  grpc_core::MutexLock lock(&socket_listeners_mu_);
+  port_listeners_.emplace_back(std::move(*single_port_listener));
```

Comment:
That is true, good catch, I hadn't noticed. The iomgr server implementation was subject to this race as well. Rather than take another mutex lock here, I opted to check again if the Listener is started after the SPL is created, then log an error about invalid usage and start the SPL if `Listener::Start` had been called concurrently.

I think there's a similar race on `Bind` and Listener destruction, I'm debating with myself whether it's worth adding a `shutdown_` flag just to check for it, I feel it will be exceedingly rare. What do you think?

Reply:
If there is an invalid usage like that the test would probably crash or TSAN could catch it? So I feel like it might be fine to not handle it right now (could be proven wrong though).

Reply:
Good question. If we had something TSAN-like on Windows, I expect we'd see something. More likely I think we would see a tiny memory leak or a crash with dangling pointers in the Bind method (or some descendent call). We aren't well instrumented to catch leaks on Windows either, but crashes are obvious.

We can write the test for the Bind-after-Start scenario if we agree that the behavior should be to return a failure, not a Crash. PosixEventEngine currently crashes. WindowsEventEngine utilizes the return Status. I'll submit a PR for that separately and see what folks think.

Reply:
https://github.com/grpc/grpc/pull/32502